File: web/src/lib/io/handlers/bundled-handler.ts
/**
 * bundled:// scheme handler. Looks up assets in a build-time table.
 *
 * The table is empty in v1 — the slot exists so akkado source like
 * `samples("bundled://default-808")` can compile, and a future build
 * step (or host registration) can populate it without changing the
 * resolver dispatch.
 */
import { FileLoadError } from '../errors';
import type { LoadOptions, LoadResult } from '../file-loader';
import type { UriHandler } from '../uri-resolver';

const BUNDLED: Map<string, ArrayBuffer> = new Map();

function strip(uri: string): string {
	if (uri.startsWith('bundled://')) return uri.slice('bundled://'.length);
	if (uri.startsWith('bundled:')) return uri.slice('bundled:'.length);
	return uri;
}

export function registerBundledAsset(name: string, bytes: ArrayBuffer): void {
	BUNDLED.set(name, bytes);
}

export const bundledHandler: UriHandler = {
	scheme: 'bundled',
	async load(uri: string, _options: LoadOptions): Promise<LoadResult> {
		const name = strip(uri);
		const data = BUNDLED.get(name);
		if (!data) {
			throw new FileLoadError('not_found', `bundled asset not found: '${name}'`);
		}
		return { data, name };
	}
};

File: cedar/src/io/handlers/bundled_handler.cpp
#include "cedar/io/handlers/bundled_handler.hpp"

#include <string>

namespace cedar {

namespace {

/// Strip "bundled://" or "bundled:" prefix if present.
std::string_view strip_bundled_scheme(std::string_view uri) {
    constexpr std::string_view triple = "bundled://";
    constexpr std::string_view single = "bundled:";
    if (uri.substr(0, triple.size()) == triple) {
        return uri.substr(triple.size());
    }
    if (uri.substr(0, single.size()) == single) {
        return uri.substr(single.size());
    }
    return uri;
}

}  // namespace

LoadResult BundledHandler::load(std::string_view uri) const {
    std::string name(strip_bundled_scheme(uri));
    auto it = assets_.find(name);
    if (it == assets_.end()) {
        return {FileLoadError{FileError::NotFound,
                              "bundled asset not found: '" + name + "'"}};
    }
    return {OwnedBuffer(std::vector<std::uint8_t>(it->second))};
}

void BundledHandler::register_asset(std::string name, std::vector<std::uint8_t> bytes) {
    assets_[std::move(name)] = std::move(bytes);
}

void BundledHandler::clear() {
    assets_.clear();
}

}  // namespace cedar
